api/src/main/java/com/flipkart/poseidon/ds/trie/Trie.java

```diff
@@ -37,73 +36,71 @@ public Trie() {
      * If no match found and reached the end then add all nodes left as firstChild
      **/
     public void add(K[] keys, V value) {
-        Boolean EXIT_STATUS = false;
-        TrieNode<K,V> currNode, prevNode;
-
-        int noOfParts = keys.length - 1;
-        int currPosition = 0;
-
-        currNode = root.firstChild;
-        prevNode = root;
-        Boolean insertAsRightSibling = false;
-
-        while (!EXIT_STATUS && currPosition <= noOfParts) {
-            if (currNode == null) {          // this means insert all parts of url as firstChild
-                for (int pos = currPosition; pos <= noOfParts; pos++) {
-                    if (!insertAsRightSibling) {
-                        currNode = prevNode.firstChild = new TrieNode<>();
-                        currNode.parent = prevNode;
-                    } else {
-                        currNode = prevNode.rightSibling = new TrieNode<>();
-                        currNode.parent = prevNode;
-                        insertAsRightSibling = false;
-                    }
-                    currNode.key = keys[pos];
+        if (root.firstChild == null) {
+            addChainAsFirstChild(root, keys, value);
+            return;
+        }
 
-                    // Mark this node as a placeholder if key is null
-                    if (currNode.key == null) {
-                        currNode.matchAny = true;
-                    }
+        TrieNode currentNode = root.firstChild;
+        TrieNode currentParent = root;
+        for (int i = 0; i < keys.length; i++) {
+            TrieNode matchingNode = findMatchingNode(currentNode, keys[i]);
+            if (matchingNode != null) {
+                currentParent = matchingNode;
+                currentNode = matchingNode.firstChild;
+                matchingNode.value = i == keys.length - 1 ? value : matchingNode.value;
+                continue;
+            }
 
-                    if (pos == noOfParts) {
-                        currNode.value = value;
-                    }
+            TrieNode newNode = new TrieNode();
+            newNode.key = keys[i];
+            newNode.matchAny = keys[i] == null;
+            newNode.value = i == keys.length - 1 ? value : null;
 
-                    // Check if new node is on right of placeHolder Node, if not move it to right
-                    if (currNode.parent.matchAny && currNode.parent.rightSibling == currNode) {
-                        currNode.parent = prevNode.parent;
-                        if (prevNode == prevNode.parent.rightSibling) {
-                            prevNode.parent.rightSibling = currNode;
-                        } else {
-                            prevNode.parent.firstChild = currNode;
-                        }
+            if (newNode.key != null) {
+                newNode.rightSibling = currentParent.firstChild;
+                currentParent.firstChild = newNode;
+            } else {
+                currentNode.rightSibling = newNode;
```

Comment:
Wildcard is not getting inserted as rightmost. This has to be fixed. We also found one more test case during get (recursive backtracking). Shrey is fixing that as well